File: TPs2/tp3/pubsInOut/tp3q13/ArvoreBi.c
/*
 * Questao 13 - Arvore Binaria em C (TP3)
 *
 * Arvore Binaria em C: Repita o exercicio anterior "Arvore Binaria",
 * realizando a implementacao em C.
 * - Chave de pesquisa: atributo nome
 * - Nao insere elemento se a chave ja estiver na arvore
 * - Entrada: IDs ate -1 (insercao), depois nomes ate FIM (pesquisa)
 * - Saida: caminho de pesquisa (raiz, esq, dir) + SIM/NAO por linha
 * - Ao final: caminhamento em ordem (inorder) de todos os registros
 * - Log: matricula_arvore_binaria.txt (matricula, comparacoes, tempo_ms)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

/* ===================== ESTRUTURAS AUXILIARES ===================== */

typedef struct Data { int ano, mes, dia; } Data;

Data parse_data(char *s) {
    Data d;
    sscanf(s, " %d - %d - %d ", &d.ano, &d.mes, &d.dia);
    return d;
}
void formatar_data(Data *d, char *buf) {
    sprintf(buf, "%02d/%02d/%04d", d->dia, d->mes, d->ano);
}

typedef struct Hora { int hora, minuto; } Hora;

Hora parse_hora(char *s) {
    Hora h;
    sscanf(s, " %d : %d ", &h.hora, &h.minuto);
    return h;
}
void formatar_hora(Hora *h, char *buf) {
    sprintf(buf, "%02d:%02d", h->hora, h->minuto);
}

typedef struct Price { int faixa; } Price;

Price parse_price(char *s) {
    Price p; p.faixa = 0;
    while (*s) { if (*s == '$') p.faixa++; s++; }
    return p;
}
void formatar_price(Price *p, char *buf) {
    int i;
    for (i = 0; i < p->faixa; i++) buf[i] = '$';
    buf[i] = '\0';
}

typedef struct TiposCozinha { char raw[200]; } TiposCozinha;

TiposCozinha create_tipos(const char *tipos) {
    TiposCozinha tc;
    int i = 0;
    while (tipos[i] != '\0') { tc.raw[i] = tipos[i]; i++; }
    tc.raw[i] = '\0';
    return tc;
}
void formatar_tipos(TiposCozinha *tc, char *buf) {
    int bi = 0;
    buf[bi++] = '[';
    for (int i = 0; tc->raw[i] != '\0'; i++)
        buf[bi++] = (tc->raw[i] == ';') ? ',' : tc->raw[i];
    buf[bi++] = ']';
    buf[bi] = '\0';
}

/* ===================== RESTAURANTE ===================== */

typedef struct Restaurante {
    int id;
    char nome[100];
    char cidade[100];
    int capacidade;
    float avaliacao;
    TiposCozinha tiposCozinha;
    Price preco;
    Data dataAbertura;
    Hora abertura, fechamento;
    bool funcionamento;
} Restaurante;

/* Localiza o '-' para separar horario de abertura/fechamento */
static int find_char(const char *s, char c) {
    for (int i = 0; s[i] != '\0'; i++)
        if (s[i] == c) return i;
    return -1;
}

static void my_trim(char *s) {
    for (int i = 0; s[i] != '\0'; i++)
        if (s[i] == '\r' || s[i] == '\n') { s[i] = '\0'; return; }
}

Restaurante parse_restaurante(char *linha) {
    Restaurante r;
    char *campos[11];
    int idx = 0;

    campos[idx++] = linha;
    for (int i = 0; linha[i] != '\0'; i++) {
        if (linha[i] == ',') {
            linha[i] = '\0';
            campos[idx++] = &linha[i + 1];
        } else if (linha[i] == '\n' || linha[i] == '\r') {
            linha[i] = '\0';
        }
    }

    sscanf(campos[0], "%d",       &r.id);
    sscanf(campos[1], "%99[^\n]", r.nome);
    sscanf(campos[2], "%99[^\n]", r.cidade);
    sscanf(campos[3], "%d",       &r.capacidade);
    sscanf(campos[4], "%f",       &r.avaliacao);

    r.tiposCozinha = create_tipos(campos[5]);
    r.preco = parse_price(campos[6]);

    int hifen = find_char(campos[7], '-');
    if (hifen >= 0) {
        campos[7][hifen] = '\0';
        r.abertura   = parse_hora(campos[7]);
        r.fechamento = parse_hora(campos[7] + hifen + 1);
    } else {
        r.abertura.hora = r.abertura.minuto = 0;
        r.fechamento.hora = r.fechamento.minuto = 0;
    }

    r.dataAbertura = parse_data(campos[8]);

    char func_str[10] = {0};
    if (campos[9] != NULL) sscanf(campos[9], "%9s", func_str);
    r.funcionamento = (strcmp(func_str, "true") == 0 || strcmp(func_str, "True") == 0);

    return r;
}

void formatar_restaurante(Restaurante *r, char *buf) {
    char dataB[20], haB[20], hfB[20], precoB[10], tiposB[200], avalB[10];
    formatar_data(&r->dataAbertura, dataB);
    formatar_hora(&r->abertura, haB);
    formatar_hora(&r->fechamento, hfB);
    formatar_price(&r->preco, precoB);
    formatar_tipos(&r->tiposCozinha, tiposB);
    sprintf(avalB, "%.1f", r->avaliacao);

    sprintf(buf,
        "[%d ## %s ## %s ## %d ## %s ## %s ## %s ## %s-%s ## %s ## %s]",
        r->id, r->nome, r->cidade, r->capacidade,
        avalB, tiposB, precoB, haB, hfB, dataB,
        r->funcionamento ? "true" : "false");
}

/* ===================== COLECAO (CSV) ===================== */

typedef struct { Restaurante array[1000]; int n; } ColecaoRestaurantes;

void ler_csv(ColecaoRestaurantes *c, const char *path) {
    FILE *f = fopen(path, "r");
    if (!f) return;
    char linha[1024];
    if (fgets(linha, sizeof(linha), f) == NULL) { fclose(f); return; }
    while (fgets(linha, sizeof(linha), f) && c->n < 1000) {
        my_trim(linha);
        c->array[c->n++] = parse_restaurante(linha);
    }
    fclose(f);
}

Restaurante *buscar_por_id(ColecaoRestaurantes *c, int id) {
    for (int i = 0; i < c->n; i++)
        if (c->array[i].id == id) return &c->array[i];
    return NULL;
}

/* ===================== ARVORE BINARIA ===================== */

/*
 * No da arvore binaria de pesquisa.
 * Armazena ponteiro para Restaurante (da colecao).
 */
typedef struct No {
    Restaurante *elemento;
    struct No *esq, *dir;
} No;

No *novo_no(Restaurante *r) {
    No *n = (No *) malloc(sizeof(No));
    n->elemento = r;
    n->esq = n->dir = NULL;
    return n;
}

/* Insercao recursiva — chave: nome */
No *inserir(No *i, Restaurante *r) {
    if (i == NULL) return novo_no(r);
    int cmp = strcmp(r->nome, i->elemento->nome);
    if (cmp < 0)      i->esq = inserir(i->esq, r);
    else if (cmp > 0) i->dir = inserir(i->dir, r);
    /* cmp == 0: elemento ja existe, nao insere */
    return i;
}

/*
 * Pesquisa recursiva.
 * Imprime o caminho (raiz esq dir ...) e SIM/NAO.
 * Acumula comparacoes em *comp.
 */
bool pesquisar(No *i, const char *nome, int *comp) {
    if (i == NULL) {
        printf(" NAO\n");
        return false;
    }
    (*comp)++;
    int c = strcmp(nome, i->elemento->nome);
    if (c == 0) {
        printf(" SIM\n");
        return true;
    } else if (c < 0) {
        printf(" esq");
        return pesquisar(i->esq, nome, comp);
    } else {
        printf(" dir");
        return pesquisar(i->dir, nome, comp);
    }
}

/* Caminhamento em ordem (inorder) — imprime em ordem alfabetica */
void caminhar_em_ordem(No *i) {
    if (i != NULL) {
        caminhar_em_ordem(i->esq);
        char buf[500];
        formatar_restaurante(i->elemento, buf);
        printf("%s\n", buf);
        caminhar_em_ordem(i->dir);
    }
}

/* ===================== MAIN ===================== */

int main(void) {
    ColecaoRestaurantes colecao;
    colecao.n = 0;
    ler_csv(&colecao, "/tmp/restaurantes.csv");

    No *raiz = NULL;
    char linha[200];

    /* Parte 1: leitura de IDs (uma por linha) ate -1, insercao na arvore */
    while (fgets(linha, sizeof(linha), stdin) != NULL) {
        my_trim(linha);
        if (strcmp(linha, "-1") == 0 || strcmp(linha, "FIM") == 0) break;
        int id;
        sscanf(linha, "%d", &id);
        Restaurante *r = buscar_por_id(&colecao, id);
        if (r != NULL) raiz = inserir(raiz, r);
    }

    int total_comp = 0;
    clock_t t_inicio = clock();

    /* Parte 2: pesquisa por nomes ate FIM */
    while (fgets(linha, sizeof(linha), stdin) != NULL) {
        my_trim(linha);
        if (strcmp(linha, "FIM") == 0 || linha[0] == '\0') break;
        printf("raiz");
        pesquisar(raiz, linha, &total_comp);
    }

    clock_t t_fim = clock();
    double tempo_ms = (double)(t_fim - t_inicio) / CLOCKS_PER_SEC * 1000.0;

    /* Parte 3: caminhamento em ordem */
    caminhar_em_ordem(raiz);

    /* Log */
    FILE *log = fopen("844387_arvore_binaria.txt", "w");
    if (log != NULL) {
        fprintf(log, "844387\t%d\t%.4f", total_comp, tempo_ms);
        fclose(log);
    }

    return 0;
}

File: TPs2/tp3/pubsInOut/tp3q12/ArvoreBi.java
import java.io.*;
import java.util.*;

// --- Classes Auxiliares de Formatação ---

class Data {
    private int ano;
    private int mes;
    private int dia;

    public Data(int ano, int mes, int dia) { 
        this.ano = ano;
        this.mes = mes;
        this.dia = dia; 
    }

    public int getAno() { return this.ano; }  
    public int getMes() { return this.mes; }  
    public int getDia() { return this.dia; }  

    public static Data parseData(String s) {
        Scanner sc = new Scanner(s);
        sc.useDelimiter("-");
        int ano = sc.nextInt();
        int mes = sc.nextInt();
        int dia = sc.nextInt();
        sc.close();
        return new Data(ano, mes, dia);
    }

    public String formatar() { 
        return String.format("%02d/%02d/%04d", dia, mes, ano); 
    }
}

class Hora {
    private int hora;
    private int minuto;

    public Hora(int hora, int minuto) { 
        this.hora = hora; 
        this.minuto = minuto; 
    }

    public static Hora parseHora(String s) {
        Scanner sc = new Scanner(s);
        sc.useDelimiter(":");
        int h = sc.nextInt();
        int m = sc.nextInt();
        sc.close();
        return new Hora(h, m);
    }

    public String formatar() { 
        return String.format("%02d:%02d", hora, minuto); 
    }
}

class Price {
    private int faixa;
    public Price(int faixa) { 
        this.faixa = faixa; 
    }

    public String formatar() {
        String resp = "";
        for (int i = 0; i < faixa; i++) {
            resp = resp + "$";
        }
        return resp;
    }
}

class TiposCozinha {
    private String[] types;

    public TiposCozinha(String[] types) { 
        this.types = types; 
    }

    public static TiposCozinha create(String tipos) {
        Scanner sc = new Scanner(tipos);
        sc.useDelimiter(";");
        int count = 0;
        for (int i = 0; i < tipos.length(); i++) {
            if (tipos.charAt(i) == ';') {
                count++;    
            }
        }
        
        String[] tps = {};
        if (sc.hasNext()) {
            tps = new String[count + 1];
            for (int i = 0; i < count + 1; i++) {
                tps[i] = sc.next();
            }
        }
        sc.close();
        return new TiposCozinha(tps);
    }

    public String formatar() {
        String res = "[";
        for (int i = 0; i < types.length; i++) {
            res = res + types[i];
            if (i < types.length - 1) {
                res = res + ",";
            }
        }
        res = res + "]";
        return res;
    }
}

// --- Classe Restaurante ---

class Restaurante {
    private int id;
    private String nome, city;
    private int capacidade;
    private double avaliacao;
    private TiposCozinha tiposCozinha;
    private int faixaPreco;
    private Data dataAbertura;
    private Hora abertura, fechamento;
    private boolean aberto;

    public Restaurante(int id, String nome, String city, int capacidade, double avaliacao,
                       TiposCozinha tiposCozinha, int faixaPreco, Data dataAbertura,
                       Hora abertura, Hora fechamento, boolean aberto) {
        this.id = id;
        this.nome = nome;
        this.city = city;
        this.capacidade = capacidade;
        this.avaliacao = avaliacao;
        this.tiposCozinha = tiposCozinha;
        this.faixaPreco = faixaPreco;
        this.dataAbertura = dataAbertura;
        this.abertura = abertura;
        this.fechamento = fechamento;
        this.aberto = aberto;
    }

    public int getId() { return id; }
    public String getNome() { return nome; }

    public static Restaurante parseRestaurante(String s) {
        Scanner sc = new Scanner(s);
        sc.useDelimiter(",");

        int id = sc.nextInt();
        String nome = sc.next();
        String city = sc.next();
        int capacidade = sc.nextInt();
        double avalia = Double.parseDouble(sc.next());
        String cozinhas = sc.next(); 
        TiposCozinha tpc = TiposCozinha.create(cozinhas); 

        int preco = sc.next().length();
        String horas = sc.next();

        Scanner scH = new Scanner(horas);
        scH.useDelimiter("-");

        Hora hAbrir = Hora.parseHora(scH.next());
        Hora hFechar = Hora.parseHora(scH.next());
        Data dAbrir = Data.parseData(sc.next());
        boolean func = sc.nextBoolean();
        
        scH.close();
        sc.close();
        return new Restaurante(id, nome, city, capacidade, avalia, tpc, preco, dAbrir, hAbrir, hFechar, func);
    }

    public String formatar() {
        Price preco = new Price(faixaPreco);
        TiposCozinha tcz = tiposCozinha;
        return String.format(Locale.US, "[%d ## %s ## %s ## %d ## %.1f ## %s ## %s ## %s-%s ## %s ## %b]",
            id, nome, city, capacidade, avaliacao,
            tcz.formatar(), preco.formatar(), 
            abertura.formatar(), fechamento.formatar(),
            dataAbertura.formatar(), aberto);
    }
}

// --- Coleção e Leitura de CSV ---

class ColecaoRestaurantes {
    private Restaurante[] rest;
    private int n;

    public ColecaoRestaurantes() {
        rest = new Restaurante[1000];
        n = 0;
    }

    public void lerCsv(String path) {
        try {
            Scanner fsc = new Scanner(new File(path));
            if (fsc.hasNextLine()) fsc.nextLine(); // Pula cabeçalho
            while (fsc.hasNextLine() && n < rest.length) {
                rest[n++] = Restaurante.parseRestaurante(fsc.nextLine());
            }
            fsc.close();
        } catch (FileNotFoundException e) {
            // Tratamento silencioso
        }
    }

    public Restaurante buscarPorId(int id) {
        for (int i = 0; i < n; i++) {
            if (rest[i].getId() == id) return rest[i];
        }
        return null;
    }
}

// --- Implementação do Nó e da Árvore Binária ---

class No {
    public Restaurante elemento;
    public No esq, dir;

    public No(Restaurante elemento) {
        this.elemento = elemento;
        this.esq = null;
        this.dir = null;
    }
}

class ArvoreBinaria {
    private No raiz;

    public ArvoreBinaria() {
        raiz = null;
    }

    public void inserir(Restaurante r) {
        raiz = inserir(raiz, r);
    }

    private No inserir(No i, Restaurante r) {
        if (i == null) {
            return new No(r);
        }
        
        int comp = r.getNome().compareTo(i.elemento.getNome());
        if (comp < 0) {
            i.esq = inserir(i.esq, r);
        } else if (comp > 0) {
            i.dir = inserir(i.dir, r);
        }
        return i;
    }

    public boolean pesquisar(String nome, int[] comp) {
        System.out.print("raiz");
        return pesquisar(raiz, nome, comp);
    }

    private boolean pesquisar(No i, String nome, int[] comp) {
        if (i == null) {
            System.out.println(" NAO");
            return false;
        }
        
        comp[0]++; 
        int c = nome.compareTo(i.elemento.getNome());
        
        if (c == 0) {
            System.out.println(" SIM");
            return true;
        } else if (c < 0) {
            System.out.print(" esq");
            return pesquisar(i.esq, nome, comp);
        } else {
            System.out.print(" dir");
            return pesquisar(i.dir, nome, comp);
        }
    }

    public void caminharEmOrdem() {
        caminharEmOrdem(raiz);
    }

    private void caminharEmOrdem(No i) {
        if (i != null) {
            caminharEmOrdem(i.esq);
            System.out.println(i.elemento.formatar());
            caminharEmOrdem(i.dir);
        }
    }
}

// --- Classe Principal / Driver Executável ---

public class ArvoreBi{
    
    public static void criarLog(int comparacoes, double tempo_ms) {
        try {
            FileWriter fw = new FileWriter("matricula_arvore_binaria.txt");
            PrintWriter pw = new PrintWriter(fw);
            pw.printf(Locale.US, "800000\t%d\t%.4f\n", comparacoes, tempo_ms);
            pw.close();
        } catch (IOException e) {
            // Ignora silenciosamente
        }
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        ColecaoRestaurantes colecao = new ColecaoRestaurantes();
        colecao.lerCsv("/tmp/restaurantes.csv");

        ArvoreBinaria arvore = new ArvoreBinaria();

        // 1ª Parte: Leitura e inserção por IDs até encontrar a flag de parada
        while (in.hasNext()) {
            String entrada = in.next();
            if (entrada.compareTo("FIM") == 0 || entrada.compareTo("-1") == 0) {
                break;
            }
            int idBusca = Integer.parseInt(entrada);
            Restaurante r = colecao.buscarPorId(idBusca);
            if (r != null) {
                arvore.inserir(r);
            }
        }

        if (in.hasNextLine()) {
            in.nextLine();
        }

        int[] totalComp = {0};
        long startTime = System.nanoTime();

        // 2ª Parte: Pesquisa por Nomes completos usando compareTo
        while (in.hasNextLine()) {
            String nomeBusca = in.nextLine().trim();
            if (nomeBusca.compareTo("FIM") == 0 || nomeBusca.isEmpty()) {
                break;
            }
            arvore.pesquisar(nomeBusca, totalComp);
        }

        long endTime = System.nanoTime();
        double tempo_ms = (endTime - startTime) / 1e6;

        // 3ª Parte: Impressão dos elementos ordenados (Caminhamento em Ordem)
        arvore.caminharEmOrdem();

        criarLog(totalComp[0], tempo_ms);
        in.close();
    }
}